### DISCUSSION TITLE
Help with CustomElementProperty

### DISCUSSION THREAD

**Discussion by ahaldane** [molstar/molstar#146] [Q&A] [answered]
Hello, 

I'm trying to get `CustomElementProperty` coloring to work for adding a custom color theme, like in basic-wrapper with the "StripedResidues" example. Before I spend more time figuring it out I wanted to check in with you: 

Is that example currently working? When I build molstar by `npm run build`, and I run the basic-example from the `build` dir with a simplehttp server, everything seems OK except for the "Apply Stripes" button, which only causes a tiny change in the ball-and-stick representation color to erase atom colorings. I expected it instead to cause a stripe color pattern in the cartoon representation. The `proteopedia` example also has a `CustomElementProperty` coloring, but it doesn't work for me either.

Related, I also would like to use `CustomElementProperty` to set the 3d labels used in the `LabelRepesentation`. It looks like that isn't currently supported, but I think I see how to modify molstar so that inside `createResidueText` it calls something like `text = theme.getLabel(l)` instead of the current hardcoded ``text = `${compId} ${authSeqId}` ``. Is that reasonable for me to try?

I think I can figure most of it out, but would appreciate a quick pointer from you that I'm on the right track. 

Thanks!

**Comment by dsehnal** [comment, marked as answer]
Hi, 

- ``CustomElementProperty`` was indeed broken if ``isApplicable`` callback wasn't present. This should now be fixed.
- If the basic functionality of this class isn't enough for you, you can look for example at https://github.com/molstar/molstar/blob/master/src/extensions/pdbe/structure-quality-report/color.ts#L44 to more advanced example.
- Yes, custom labels from a custom property indeed need to be implemented "by hand".

This should get you started:

```ts
export const MyLabels3D = CreateTransformer({
    name: 'my-labels-3d',
    display: 'My Labels',
    from: PluginStateObject.Molecule.Structure,
    to: PluginStateObject.Shape.Representation3D,
    params: {
        // custom prams
        ...LabelParams,
    },
})({
    apply({ a, params }, plugin: PluginContext) {
        return Task.create('Labels', async (ctx) => {
            const data: LabelData = <create the object>; // access the custom property or pass the labels as params of the transform
            
            // createIndexLabelData(a.data, params.selection);
            const repr = LabelRepresentation(
                {
                    webgl: plugin.canvas3d?.webgl,
                    ...plugin.representation.structure.themes,
                },
                () => LabelParams
            );
            await repr.createOrUpdate(params, data).runInContext(ctx);
            repr.setState({ pickable: false });
            return new PluginStateObject.Shape.Representation3D({ repr, source: a }, { label: `Label` });
        });
    },
    update({ a, b, newParams }) {
        return Task.create('Labels', async (ctx) => {
            const data: LabelData = <updated labels>;
            await b.data.repr.createOrUpdate(newParams, data).runInContext(ctx);
            return StateTransformer.UpdateResult.Updated;
        });
    },
});

/// .....

await this.plugin.build().to(structureSelector).apply(MyLabels3D).commit();
```

**Reply by ahaldane** [reply]
Thanks, the custom colors work now. I'll take a look at the transform you suggest too, thanks.
